SwiftCompilerSources/Sources/Optimizer/Utilities/AccessUtils.swift

```diff
@@ -0,0 +1,435 @@
+    return answer(nil)
+  }
+  
+  mutating func walkUp(address: Value, path: Path, state: State) -> WalkResult {
+    if address is IndexAddrInst {
+      // Track that we crossed an `index_addr` during the walk-up
+      return walkUpDefault(address: address, path: path, state: State(cast: state.cast, indexAddr: true))
+    } else if state.indexAddr && !canBeOperandOfIndexAddr(address) {
+      // An `index_addr` instruction cannot be derived from an address
+      // projection. Bail out
+      self.result = nil
+      return .abortWalk
+    } else if let ba = address as? BeginAccessInst, _beginAccess == nil {
+      _beginAccess = ba
+    }
+    return walkUpDefault(address: address, path: path, state: State(cast: state.cast, indexAddr: false))
+  }
+  
+  mutating func rootDef(value: Value, path: Path, state: State) -> WalkResult { fatalError("Unreachable") }
+  
+  mutating func walkUp(value: Value, path: Path, state: State) -> WalkResult {
+    assert(state.cast != nil, "A value walkUp was requested but no cast provided as state")
+    switch value {
+    case is FunctionArgument:
+      // TODO: maybe define `walkUp(value:` to track certain crossed
+      // instructions like `RefToBridgeObjectInst` (?)
+      self.result = (value, .Pointer, path)
+      return .continueWalk
+    case let atp as AddressToPointerInst:
+      if let cast = state.cast, cast.type == atp.operand.type {
+        return walkUp(address: atp.operand, path: path, state: State(cast: nil, indexAddr: false))
+      }
+    case let arg as BlockArgument:
+      return walkUpDefault(value: arg, path: path, state: state)
+    default:
+      break
+    }
+    
+    self.result = nil
+    return .abortWalk
+  }
+}
+
+/// A walker utility that given an address definition computes the `AccessPath`
+/// of the access (the base address and the address projections to the accessed fields) and
+/// the innermost enclosing scope (`begin_access`).
+struct AccessPathWalker {
+  private var walker = AccessPathWalkerInternal()
+  mutating func compute(_ address: Value) -> AccessPath? {
+    assert(address.type.isAddress, "Expected address")
+    walker.start()
+    _ = walker.walkUp(address: address, path: AccessPathWalkerInternal.Path(),
+                      state: AccessPathWalkerInternal.State(cast: nil, indexAddr: false))
+    
+    if let (instr, kind, path) = walker.result {
+      return AccessPath(base: AccessBase(baseAddress: instr, storageKind: kind), projectionPath: path)
+    }
+    
+    return nil
+  }
+  
+  mutating func pathWithScope(of address: Value) -> (AccessPath?, EnclosingScope?) {
+    let ap = compute(address)
+    return (ap, walker.scope)
+  }
+  
+  mutating func base(of address: Value) -> AccessBase? {
+    compute(address)?.base
+  }
+  
+  mutating func scope(of address: Value) -> EnclosingScope {
+    _ = compute(address)
+    return walker.scope!
+  }
+}
+
+struct AccessRootWalkerInternal : ValueUseDefWalker, WithDefaultCache {
```

Comment:
> AccessPathWalkerInternal : AddressUseDefWalker, ValueUseDefWalker

It's really unfortunate that AccessPathWalkerInternal conforms to ValueUseDefWalker. It pulls in all the implementation used to walk values, which doesn't normally make sense for computing an access path. There's no way to tell anymore which operations may part of an access path. This partially defeats the purpose of separating the address and value walkers.

I think this was done to allow sequences like:
```
address_to_pointer
struct
struct_extract
pointer_to_address
```

But the operations on the pointer value don't contribute to the address projection path.  And the set of allowable operations on pointer types are completely different from those for addresses or references (non-trivial types).

I originally thought that multiple walkers would be involved in a single SSA traversal and they would just share a context. That makes sense to me because the pointer and address walkers should neither share the same state nor share any implementation. The pointer walker doesn't have anything to do with the access base or access path. None of the operations overlap.

I would define a completely separate pointer walker. You can assert that the pointer path is empty when it calls back to the parent walker. The AccessPathWalker would not need to deal with any value walking stuff.